rollup/circuitcapacitychecker/impl.go

```diff
@@ -0,0 +1,117 @@
+//go:build circuit_capacity_checker
+
+package circuitcapacitychecker
+
+/*
+#cgo LDFLAGS: -lm -ldl -lzkp -lzktrie
+#include <stdlib.h>
+#include "./libzkp/libzkp.h"
+*/
+import "C" //nolint:typecheck
+
+import (
+	"encoding/json"
+	"sync"
+	"unsafe"
+
+	"github.com/scroll-tech/go-ethereum/core/types"
+	"github.com/scroll-tech/go-ethereum/log"
+)
+
+func init() {
+	C.init()
+}
+
+type CircuitCapacityChecker struct {
+	sync.Mutex
+	id uint64
+}
+
+func NewCircuitCapacityChecker() *CircuitCapacityChecker {
+	id := C.new_circuit_capacity_checker()
```

Comment:
I think this needs a lock somewhere

Reply:
fixed in https://github.com/scroll-tech/go-ethereum/pull/321/commits/baf2eccf12d9bcf16a0f63ed7beeff7a7169f162